File: Sliding Window/0632.Smallest Range Covering Elements from K Lists/Solution.py
class Solution:
    def smallestRange(self, nums: List[List[int]]) -> List[int]:
        q = []
        for i, row in enumerate(nums):
            for v in row:
                q.append((v, i))
        q.sort()
        cnt = dict()
        l = r = 0
        ans = [q[0][0], q[-1][0]]
        while r < len(q):
            cnt[q[r][1]] = cnt.get(q[r][1], 0) + 1
            while len(cnt) == len(nums):
                if ans[1] - ans[0] > q[r][0] - q[l][0]:
                    ans = [q[l][0], q[r][0]]
                cnt[q[l][1]] -= 1
                if cnt[q[l][1]] == 0:
                    cnt.pop(q[l][1])
                l += 1
            r += 1
        return ans

File: Sliding Window/0632.Smallest Range Covering Elements from K Lists/Solution.java
class Solution {
    public int[] smallestRange(List<List<Integer>> nums) {
        int n = nums.size();
        List<int[]> q = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int v : nums.get(i)) {
                q.add(new int[]{v, i});
            }
        }
        Collections.sort(q, Comparator.comparingInt(o -> o[0]));
        Map<Integer, Integer> cnt = new HashMap<>();
        int l = 0, r = 0, ansL = q.get(0)[0], ansR = q.get(q.size() - 1)[0];
        while (r < q.size()) {
            int[] cur = q.get(r);
            cnt.put(cur[1], cnt.getOrDefault(cur[1], 0) + 1);
            while (cnt.size() == n) {
                if (q.get(r)[0] - q.get(l)[0] < ansR - ansL) {
                    ansL = q.get(l)[0];
                    ansR = q.get(r)[0];
                }
                int[] curL = q.get(l++);
                cnt.put(curL[1], cnt.get(curL[1]) - 1);
                if (cnt.get(curL[1]) == 0) {
                    cnt.remove(curL[1]);
                }
            }
            r++;
        }
        return new int[]{ansL, ansR};
    }
}
